File: app/core/src/main/java/stirling/software/SPDF/service/pdf/OcrContentRegion.java
package stirling.software.SPDF.service.pdf;

/** Vertical OCR bounds as ratios of page height (0 = top, 1 = bottom). */
public record OcrContentRegion(float topRatio, float bottomRatio) {

    private static final float MIN_BAND = 0.05f;

    public static OcrContentRegion fullPage() {
        return new OcrContentRegion(0f, 1f);
    }

    public static OcrContentRegion defaultsForBooks() {
        return new OcrContentRegion(0.06f, 0.87f);
    }

    public static OcrContentRegion parse(Float topRatio, Float bottomRatio) {
        if (topRatio == null && bottomRatio == null) {
            return fullPage();
        }
        return new OcrContentRegion(
                        topRatio != null ? topRatio : 0f, bottomRatio != null ? bottomRatio : 1f)
                .normalized();
    }

    public OcrContentRegion normalized() {
        float top = clamp(topRatio, 0f, 0.45f);
        float bottom = clamp(bottomRatio, top + MIN_BAND, 1f);
        return new OcrContentRegion(top, bottom);
    }

    public boolean isFullPage() {
        return topRatio <= 0.001f && bottomRatio >= 0.999f;
    }

    private static float clamp(float value, float min, float max) {
        return Math.max(min, Math.min(max, value));
    }
}

File: frontend/electron/shared/ocrContentRegion.ts
/** Vertical OCR bounds as ratios of page height (0 = top, 1 = bottom). */
export interface OcrContentRegion {
  contentTop: number;
  contentBottom: number;
}

export const OCR_CONTENT_REGION_MIN_BAND = 0.05;

export const DEFAULT_OCR_CONTENT_REGION: OcrContentRegion = {
  contentTop: 0.06,
  contentBottom: 0.87,
};

export function normalizeOcrContentRegion(
  value: Partial<OcrContentRegion> | null | undefined,
): OcrContentRegion {
  const top =
    typeof value?.contentTop === "number" && Number.isFinite(value.contentTop)
      ? value.contentTop
      : DEFAULT_OCR_CONTENT_REGION.contentTop;
  const bottom =
    typeof value?.contentBottom === "number" && Number.isFinite(value.contentBottom)
      ? value.contentBottom
      : DEFAULT_OCR_CONTENT_REGION.contentBottom;
  return clampOcrContentRegion({ contentTop: top, contentBottom: bottom });
}

export function clampOcrContentRegion(region: OcrContentRegion): OcrContentRegion {
  const top = Math.max(0, Math.min(region.contentTop, 0.45));
  const bottom = Math.max(top + OCR_CONTENT_REGION_MIN_BAND, Math.min(region.contentBottom, 1));
  return { contentTop: top, contentBottom: bottom };
}

export function isFullPageRegion(region: OcrContentRegion): boolean {
  return region.contentTop <= 0.001 && region.contentBottom >= 0.999;
}
